### DISCUSSION TITLE
Initial MVP

### DISCUSSION THREAD

**Discussion by dinophile** [cds-snc/url-shortener#29] [Announcements]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members and plan out features.

**Comment by dinophile** [comment]
@patheard asked:

Do we need to watch out for random string collision if we're expiring links?  Case I'm thinking of is:

1. Link is created and published.
2. Link expires and a new link is created that ends up using the same random string.
3. Published content now points to the wrong link.

This is almost certainly an edge case.

**Reply by sylviamclaughlin** [reply]
@patheard,

This is a good point and we should certainly think about this. Right now the string generated is 8 characters (it can be lower case, upper case and symbols) and the encoded string that is taken is {url}{timestamp}, so if my math is right the number is 3.026×10 to the power of 15 unique combinations. Obviously if we extend the length the number will be higher. Event though small likelihood, we should still think about this since it could potentially happen so we could keep track of all generated shortlinks.

**Comment by dinophile** [comment]
@patheard asked:

As this allow list changes, should we have a way to invalidate links that were previously allowed but have been removed from the allow list?

**Reply by sylviamclaughlin** [reply]
That's a good idea. We can build in some functionality to traverse the links if the allow list changes.

**Comment by dinophile** [comment]
@patheard asked:

Is Pavithra also working as a dev on this?

**Reply by dinophile** [reply]
Woo I can answer this one: Yes that was the intention!

**Comment by dinophile** [comment]
@jeberhardt suggested:

I think the "random string" should have some restrictions on it, to make it more readable / user friendly. For example: no repeating characters, excluding similar characters.

**Reply by sylviamclaughlin** [reply]
We could, it will complicate the algorithm if we build that in but is totally doable. We will just have to provide a list of similar characters and build in strict rules on how it should be more readable/user friendly.

**Comment by dinophile** [comment]
@patheard asked:

Will this include things like an audit trail around things like who created and edited a shortened URL?

**Reply by dinophile** [reply]
In response @jeberhardt asked:

Once created, can a shortened URL be edited?

**Reply by dinophile** [reply]
@patheard replied:

Maybe?  The case I'm thinking of is someone made a typo and they'd like to fix it without having to republish the message.

**Reply by dinophile** [reply]
The way we solved this as a first pass at ontario.ca was to just allow them to delete the original URL and try again. That way no PUTS/updates needed to be managed. (Just a good old fashioned delete...ahhh....)

But that was with an account managed system.

We'll have to have a way to store and connect the user who created the URL to either delete it (first pass) or edit the domain ONLY. Not the random string.

**Reply by sylviamclaughlin** [reply]
I think for that we will have to an account managed system since we would need to know who the user is otherwise there is no way to tell. We will have to build something like an admin menu.

**Comment by sylviamclaughlin** [comment]
Have we confirmed/decided if this is going to have user authentication and login?

**Comment by sylviamclaughlin** [comment]
Is the initial MVP just the API or the frontend as well?

**Reply by jeberhardt** [reply]
I think there should be a frontend as well, if for no other reason than to ensure we can test it out.

**Comment by sylviamclaughlin** [comment]
If the Initial MVP has a frontend, should we keep Jinja as is the current templating language or should we build a separate frontend using something like React?

**Reply by jeberhardt** [reply]
I would suggest that we use React. The front-end can be totally separate from the API. Both `articles` and `forms` use React. It would be good to have consistency across projects.
